pkg/controllermanager/controller/shoot/maintenance/reconciler.go

```diff
@@ -93,99 +96,151 @@ func requeueAfterDuration(shoot *gardencorev1beta1.Shoot) (time.Duration, time.T
-	for i, w := range shoot.Spec.Provider.Workers {
+	for i, w := range maintainedShoot.Spec.Provider.Workers {
 		if w.Kubernetes == nil || w.Kubernetes.Version == nil {
 			continue
 		}
 
 		workerLog := log.WithValues("worker", w.Name)
-
-		reasonForWorkerPoolKubernetesUpdate, err := maintainKubernetesVersion(workerLog, *w.Kubernetes.Version, shoot.Spec.Maintenance.AutoUpdate.KubernetesVersion, cloudProfile, func(v string) error {
+		name := "Worker Pool " + w.Name
+		reasonForWorkerPoolKubernetesUpdate, err := maintainKubernetesVersion(workerLog, *w.Kubernetes.Version, maintainedShoot.Spec.Maintenance.AutoUpdate.KubernetesVersion, cloudProfile, func(v string) error {
 			workerPoolSemver, err := semver.NewVersion(v)
 			if err != nil {
 				return err
 			}
 			// If during autoupdate a worker pool kubernetes gets forcefully updated to the next minor which might be higher than the same minor of the shoot, take this
-			if workerPoolSemver.GreaterThan(shootSemver) {
-				workerPoolSemver = shootSemver
+			if workerPoolSemver.GreaterThan(shootKubernetesVersion) {
+				workerPoolSemver = shootKubernetesVersion
 			}
 			v = workerPoolSemver.String()
-			shoot.Spec.Provider.Workers[i].Kubernetes.Version = &v
+			maintainedShoot.Spec.Provider.Workers[i].Kubernetes.Version = &v
 			return nil
-		})
+		}, name)
 		if err != nil {
 			// continue execution to allow the machine image version update
 			workerLog.Error(err, "Could not maintain kubernetes version for worker pool")
 		}
-		reasonsForWorkerPoolKubernetesUpdate[w.Name] = reasonForWorkerPoolKubernetesUpdate
+
+		reasonForKubernetesUpdate = append(reasonForKubernetesUpdate, reasonForWorkerPoolKubernetesUpdate...)
+	}
+	operations = append(operations, reasonForKubernetesUpdate...)
+
+	operation := maintainOperation(maintainedShoot)
+	if operation != "" {
+		operations = append(operations, fmt.Sprintf("Added %q operation annotation", operation))
+	}
+
+	if len(operations) > 0 {
+		patch := client.MergeFrom(shoot.DeepCopy())
+		shoot.Status.LastMaintenance = &gardencorev1beta1.LastMaintenance{
+			Description:   strings.Join(operations, ", "),
+			TriggeredTime: metav1.Time{Time: r.Clock.Now()},
+			State:         gardencorev1beta1.LastOperationStateProcessing,
+		}
+
+		// First dry run the update call to check if it can be executed successfully.
+		// If not shoot maintenance is marked as failed and is retried only in
+		// next maintenance window.
+		if err := r.Client.Update(ctx, maintainedShoot, &client.UpdateOptions{
+			DryRun: []string{metav1.DryRunAll},
+		}); err != nil {
+			// If shoot maintenance is triggerd by `gardener.cloud/operation=maintain` annotation and if it fails in dry run
+			// maintain operation annotation need to be removed so that if reason for failure is fixed and maintenance is triggerd
+			// again using maintain operation annotation then it should not fail with the reason that annotation is already present
```

Comment:
```suggestion
			// If shoot maintenance is triggered by `gardener.cloud/operation=maintain` annotation and if it fails in dry run,
			// then 'maintain' operation annotation needs to be removed so that if reason for failure is fixed and maintenance is triggered
			// again via 'maintain' operation annotation then it should not fail with the reason that annotation is already present
```

I'm missing some explanation that we are using the `shoots/status` subresource here but can still remove the annotation (even though this is in `metadata` and not `status`).

Reply:
Yes, here on status update only `ShootSpec` is kept in its original form.
https://github.com/gardener/gardener/blob/a2f7de0badaae6170d7b9b84c163b8cab43a84d2/pkg/registry/core/shoot/strategy.go#L258-L267

Reply:
Yes, but I asked you to enhance this comment here because it was not obvious (at least to me) that this is possible (updating metadata with `status` subresource).

Reply:
Done PTAL.